ovs/dal/datalist.py

```diff
@@ -103,16 +116,132 @@ def guids(self):
+        :rtype: NoneType
+        """
+        # Force query to rerun
+        self._executed = False
+        self._guids = None
+        self._data = {}
+        self._objects = {}
+        # Reset index information
+        self.from_index = 'none'
+        # Reset caching info
+        self._can_cache = True
+
+    def set_query(self, query):
+        """
+        Sets the query to apply to a different query
+        :param query: The query to perform. If this query is different from the previous one, the previously cached result will be replaced
+        with results of this query. If the query is identical and the result is cached, it will return the cached result
+        If None is supplied, a default query which does not filter anything will be set
+        :type query: dict or NoneType
+        Definitions:
+        * <query>: Should be a dictionary:
+                   {'type' : DataList.where_operator.XYZ,
+                    'items': <items>}
+        * <filter>: A tuple defining a single expression:
+                    (<field>, DataList.operator.XYZ, <value> [, <ignore_case>])
+                    The field is any property you would also find on the given object. In case of
+                    properties, you can dot as far as you like.
+        * <items>: A list of one or more <query> or <filter> items. This means the query structure is recursive and
+                   complex queries are possible
+        :return: None
+        :rtype: NoneType
+        """
+        if query is None:
+            query = {'type': DataList.where_operator.AND, 'items': []}
+        self._validate_query(query)
+        self._query = query
+        if self._provided_key is True:
+            # Cache has to be reset as it is no longer valid
+            self._volatile.delete(self._key)
+        else:
+            self.set_key()
+        self._reset_list()
+
+    def set_guids(self, guids):
+        """
+        Sets up a list of guids to apply the query too
+        :param guids: List of guids to query on or None in case you wish to query all items again
+        :type guids: list[basestring] or NoneType
+        :return: None
+        :rtype: NoneType
+        """
+        if guids is not None:
+            self._validate_guids(guids)
+        self._provided_guids = guids
+        self._provided_keys = None
+        if self._provided_key is True:
+            # Cache has to be reset as it is no longer valid
+            self._volatile.delete(self._key)
+        else:
+            self.set_key()
+        self._reset_list()
+
+    @staticmethod
+    def _validate_query(query):
+        """
+        Validates if a query is of the format we'd expect
+        :param query: Query to perform
+        :type query: dict
+        :return: None
+        :rtype: NoneType
+        :raises: ValueError if the query is not valid
+        """
+        if query is not None and not isinstance(query, dict) and not all(k in query for k in ("type", "items")):
```

Comment:
You'll never reach the final statement (`not all(k in query for k in ("type", "items"))`) in case you have a dict
Which means you'll never check the keys in the dict

Perhaps a small unitttest to verify this?